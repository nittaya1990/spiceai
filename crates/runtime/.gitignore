abc-sqlite.db-wal
abc-sqlite.db-shm
