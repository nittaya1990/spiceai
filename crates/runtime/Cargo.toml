[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
