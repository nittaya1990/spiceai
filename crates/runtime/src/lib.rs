/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#![allow(clippy::missing_errors_doc)]

use std::net::SocketAddr;
use std::{collections::HashMap, sync::Arc};

use crate::{
    auth::EndpointAuth, dataconnector::DataConnector, datafusion::DataFusion,
    internal_table::Error as InternalTableError, model::ENABLE_MODEL_SUPPORT_MESSAGE,
};
use ::datafusion::error::DataFusionError;
use ::datafusion::sql::{sqlparser, TableReference};
use app::App;
use builder::RuntimeBuilder;
use config::Config;
use dataconnector::ConnectorComponent;
use datasets_health_monitor::DatasetsHealthMonitor;
use extension::ExtensionFactory;
use flight::RateLimits;
#[cfg(feature = "openapi")]
pub use http::ApiDoc;
use model::{EmbeddingModelStore, EvalScorerRegistry, LLMModelStore};

use model_components::model::Model;
pub use notify::Error as NotifyError;
use secrecy::SecretString;
use secrets::{ParamStr, Secrets};
use snafu::prelude::*;
use spicepod::component::eval::Eval;
use status::ComponentStatus;
use tls::TlsConfig;
use tokio::sync::{oneshot::error::RecvError, RwLock};
use tools::factory::default_available_catalogs;
use tools::{catalog::SpiceToolCatalog, Tooling};
pub use util::shutdown_signal;

use crate::extension::Extension;
pub mod accelerated_table;
pub mod auth;
mod builder;
pub mod catalogconnector;
pub mod component;
pub mod config;
pub mod dataaccelerator;
pub mod dataconnector;
pub mod datafusion;
pub mod datasets_health_monitor;
pub mod dataupdate;
pub mod embeddings;
pub mod execution_plan;
pub mod extension;
pub mod federated_table;
pub mod flight;
mod http;
mod init;
pub mod internal_table;
mod metrics;
mod metrics_server;
pub mod model;
pub mod object_store_registry;
pub mod objectstore;
mod opentelemetry;
mod parameters;
pub mod podswatcher;
pub mod request;
pub mod secrets;
pub mod spice_metrics;
pub mod status;
pub mod task_history;
pub mod timing;
pub mod tls;
pub mod tools;
pub mod topological_ordering;
pub(crate) mod tracers;
mod tracing_util;
mod view;

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display("Unable to start HTTP server: {source}"))]
    UnableToStartHttpServer { source: http::Error },

    #[snafu(display("{source}"))]
    UnableToJoinTask { source: tokio::task::JoinError },

    #[snafu(display("Unable to start Prometheus metrics server: {source}"))]
    UnableToStartMetricsServer { source: metrics_server::Error },

    #[snafu(display("Unable to start Flight server: {source}"))]
    UnableToStartFlightServer { source: flight::Error },

    #[snafu(display("Unable to start OpenTelemetry server: {source}"))]
    UnableToStartOpenTelemetryServer { source: opentelemetry::Error },

    #[snafu(display("Unknown data source: {data_source}"))]
    UnknownDataSource { data_source: String },

    #[snafu(display("Unable to create data backend: {source}"))]
    UnableToCreateBackend { source: datafusion::Error },

    #[snafu(display("Unable to attach view: {source}"))]
    UnableToAttachView { source: datafusion::Error },

    #[snafu(display("Failed to start pods watcher: {source}"))]
    UnableToInitializePodsWatcher { source: NotifyError },

    #[snafu(display("{source}"))]
    UnableToInitializeDataConnector {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("{source}"))]
    UnableToInitializeCatalogConnector {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("{source}"))]
    UnableToInitializeLlm {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("{source}"))]
    UnableToInitializeEmbeddingModel {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("{source}"))]
    UnableToInitializeLlmTool {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("Unknown data connector: {data_connector}.\nSpecify a valid data connector and retry. For details, visit: https://spiceai.org/docs/components/data-connectors"))]
    UnknownDataConnector { data_connector: String },

    #[snafu(display("Unknown catalog connector: {catalog_connector}.\nSpecify a valid catalog connector and retry. For details, visit: https://spiceai.org/docs/components/catalogs"))]
    UnknownCatalogConnector { catalog_connector: String },

    #[snafu(display("The runtime is built without ODBC support.\nBuild Spice.ai OSS with the `odbc` feature enabled or use the Docker image that includes ODBC support.\nFor details, visit: https://spiceai.org/docs/components/data-connectors/odbc"))]
    OdbcNotInstalled,

    #[snafu(display("Unable to load secrets for data connector: {data_connector}"))]
    UnableToLoadDataConnectorSecrets { data_connector: String },

    #[snafu(display("Unable to get secret for data connector {data_connector}: {source}"))]
    UnableToGetSecretForDataConnector {
        source: Box<dyn std::error::Error + Send + Sync>,
        data_connector: String,
    },

    #[snafu(display("Unable to get secret for LLM: {source}"))]
    UnableToGetSecretForLLM {
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("Failed to setup the {connector_component} ({data_connector}).\n{source}"))]
    UnableToAttachDataConnector {
        source: datafusion::Error,
        connector_component: ConnectorComponent,
        data_connector: String,
    },

    #[snafu(display("Unable to load SQL file {file}: {source}"))]
    UnableToLoadSqlFile {
        file: String,
        source: std::io::Error,
    },

    #[snafu(display("Unable to parse SQL: {source}"))]
    UnableToParseSql {
        source: sqlparser::parser::ParserError,
    },

    #[snafu(display("Unable to create view: {reason}"))]
    UnableToCreateView { reason: String },

    #[snafu(display("Specify the SQL string for view {name} using either `sql: SELECT * FROM...` inline or as a file reference with `sql_ref: my_view.sql`"))]
    NeedToSpecifySQLView { name: String },

    #[snafu(display("An accelerated table was configured as read_write without setting replication.enabled = true"))]
    AcceleratedReadWriteTableWithoutReplication,

    #[snafu(display("An accelerated table for {dataset_name} was configured with 'refresh_mode = changes', but the data connector doesn't support a changes stream."))]
    AcceleratedTableInvalidChanges { dataset_name: String },

    #[snafu(display("Expected acceleration settings for {name}, found None"))]
    ExpectedAccelerationSettings { name: String },

    #[snafu(display("The accelerator engine {name} is not available. Valid engines are arrow, duckdb, sqlite, and postgres."))]
    AcceleratorEngineNotAvailable { name: String },

    #[snafu(display("The accelerator engine {name} failed to initialize: {source}"))]
    AcceleratorInitializationFailed {
        name: String,
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display(
        "Dataset names should not include a catalog. Unexpected '{}' in '{}'. Remove '{}' from the dataset name and try again.",
        catalog,
        name,
        catalog,
    ))]
    DatasetNameIncludesCatalog { catalog: Arc<str>, name: Arc<str> },

    #[snafu(display("Unable to load dataset connector: {dataset}"))]
    UnableToLoadDatasetConnector { dataset: TableReference },

    #[snafu(display("Unable to load data connector for catalog {catalog}: {source}"))]
    UnableToLoadCatalogConnector {
        catalog: String,
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[snafu(display("The data connector {dataconnector} doesn't support catalogs."))]
    DataConnectorDoesntSupportCatalogs { dataconnector: String },

    #[snafu(display("Unable to create accelerated table: {dataset}, {source}"))]
    UnableToCreateAcceleratedTable {
        dataset: TableReference,
        source: datafusion::Error,
    },

    #[snafu(display("Unable to receive accelerated table status: {source}"))]
    UnableToReceiveAcceleratedTableStatus { source: RecvError },

    #[snafu(display("Unable to start local metrics: {source}"))]
    UnableToStartLocalMetrics { source: spice_metrics::Error },

    #[snafu(display("Unable to track task history: {source}"))]
    UnableToTrackTaskHistory { source: task_history::Error },

    #[snafu(display("Unable to create metrics table: {source}"))]
    UnableToCreateMetricsTable { source: DataFusionError },

    #[snafu(display("Unable to create eval runs table: {source}"))]
    UnableToCreateEvalRunsTable { source: InternalTableError },

    #[snafu(display("Unable to register metrics table: {source}"))]
    UnableToRegisterMetricsTable { source: datafusion::Error },

    #[snafu(display("Invalid dataset defined in Spicepod: {source}"))]
    InvalidSpicepodDataset {
        source: crate::component::dataset::Error,
    },

    #[snafu(display("Invalid glob pattern {pattern}: {source}"))]
    InvalidGlobPattern {
        pattern: String,
        source: globset::Error,
    },

    #[snafu(display("Error converting GlobSet to Regex: {source}"))]
    ErrorConvertingGlobSetToRegex { source: globset::Error },

    #[snafu(display("Unable to create directory: {source}"))]
    UnableToCreateDirectory { source: std::io::Error },

    #[snafu(display("{source}"))]
    ComponentError { source: component::Error },
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(Clone, Copy)]
pub struct LogErrors(pub bool);

#[derive(Clone)]
pub struct Runtime {
    app: Arc<RwLock<Option<Arc<App>>>>,
    df: Arc<DataFusion>,
    models: Arc<RwLock<HashMap<String, Model>>>,
    llms: Arc<RwLock<LLMModelStore>>,
    embeds: Arc<RwLock<EmbeddingModelStore>>,
    tools: Arc<RwLock<HashMap<String, Tooling>>>,
    evals: Arc<RwLock<Vec<Eval>>>,
    eval_scorers: EvalScorerRegistry,
    pods_watcher: Arc<RwLock<Option<podswatcher::PodsWatcher>>>,
    secrets: Arc<RwLock<secrets::Secrets>>,
    datasets_health_monitor: Option<Arc<DatasetsHealthMonitor>>,
    metrics_endpoint: Option<SocketAddr>,
    prometheus_registry: Option<prometheus::Registry>,
    rate_limits: Arc<RateLimits>,

    autoload_extensions: Arc<HashMap<String, Box<dyn ExtensionFactory>>>,
    extensions: Arc<RwLock<HashMap<String, Arc<dyn Extension>>>>,
    spaced_tracer: Arc<tracers::SpacedTracer>,

    status: Arc<status::RuntimeStatus>,
}

impl Runtime {
    #[must_use]
    pub fn builder() -> RuntimeBuilder {
        RuntimeBuilder::new()
    }

    #[must_use]
    pub fn datafusion(&self) -> Arc<DataFusion> {
        Arc::clone(&self.df)
    }

    #[must_use]
    pub fn secrets(&self) -> Arc<RwLock<secrets::Secrets>> {
        Arc::clone(&self.secrets)
    }

    #[must_use]
    pub fn status(&self) -> Arc<status::RuntimeStatus> {
        Arc::clone(&self.status)
    }

    #[must_use]
    pub fn embeds(&self) -> Arc<RwLock<EmbeddingModelStore>> {
        Arc::clone(&self.embeds)
    }

    #[must_use]
    pub fn app(&self) -> Arc<RwLock<Option<Arc<App>>>> {
        Arc::clone(&self.app)
    }

    /// Requests a loaded extension, or will attempt to load it if part of the autoloaded extensions.
    pub async fn extension(&self, name: &str) -> Option<Arc<dyn Extension>> {
        let extensions = self.extensions.read().await;

        if let Some(extension) = extensions.get(name) {
            return Some(Arc::clone(extension));
        }
        drop(extensions);

        if let Some(autoload_factory) = self.autoload_extensions.get(name) {
            let mut extensions = self.extensions.write().await;
            let mut extension = autoload_factory.create();
            let extension_name = extension.name().to_string();
            if let Err(err) = extension.initialize(self).await {
                tracing::error!("Unable to initialize extension {extension_name}: {err}");
                return None;
            }

            if let Err(err) = extension.on_start(self).await {
                tracing::error!("Unable to start extension {extension_name}: {err}");
                return None;
            }

            extensions.insert(extension_name.clone(), extension.into());
            return extensions.get(&extension_name).cloned();
        }

        None
    }

    /// Starts the HTTP, Flight, OpenTelemetry and Metrics servers all listening on the ports specified in the given `Config`.
    ///
    /// The future returned by this function drives the individual server futures and will only return once the servers are shutdown.
    ///
    /// It is recommended to start the servers in parallel to loading the Runtime components to speed up startup.
    pub async fn start_servers(
        self: Arc<Self>,
        config: Config,
        tls_config: Option<Arc<TlsConfig>>,
        endpoint_auth: EndpointAuth,
    ) -> Result<()> {
        self.register_metrics_table(self.prometheus_registry.is_some())
            .await?;

        let http_auth = endpoint_auth.http_auth.clone();
        let http_server_future = tokio::spawn(http::start(
            config.http_bind_address,
            Arc::clone(&self),
            config.clone().into(),
            tls_config.clone(),
            http_auth,
        ));

        // Spawn the metrics server in the background
        let metrics_endpoint = self.metrics_endpoint;
        let prometheus_registry = self.prometheus_registry.clone();
        let cloned_tls_config = tls_config.clone();
        tokio::spawn(async move {
            if let Err(e) =
                metrics_server::start(metrics_endpoint, prometheus_registry, cloned_tls_config)
                    .await
            {
                tracing::error!("Prometheus metrics server error: {e}");
            }
        });

        let flight_server_future = tokio::spawn(flight::start(
            config.flight_bind_address,
            self.app.read().await.as_ref().map(Arc::clone),
            Arc::clone(&self.df),
            tls_config.clone(),
            endpoint_auth.clone(),
            Arc::clone(&self.rate_limits),
        ));
        let open_telemetry_server_future = tokio::spawn(opentelemetry::start(
            config.open_telemetry_bind_address,
            Arc::clone(&self.df),
            tls_config.clone(),
            endpoint_auth.grpc_auth.clone(),
        ));

        let pods_watcher_future = if self.pods_watcher.read().await.is_some() {
            Some(self.start_pods_watcher())
        } else {
            None
        };

        if let Some(tls_config) = tls_config {
            match tls_config.subject_name() {
                Some(subject_name) => {
                    tracing::info!("Endpoints secured with TLS using certificate: {subject_name}");
                }
                None => {
                    tracing::info!("Endpoints secured with TLS");
                }
            }
        }

        tokio::select! {
            http_res = http_server_future => {
                match http_res {
                    Ok(http_res) => http_res.context(UnableToStartHttpServerSnafu),
                    Err(source) => {
                        Err(Error::UnableToJoinTask { source })
                    }
                }
             },
            flight_res = flight_server_future => {
                match flight_res {
                    Ok(flight_res) => flight_res.context(UnableToStartFlightServerSnafu),
                    Err(source) => {
                        Err(Error::UnableToJoinTask { source })
                    }
                }
            },
            open_telemetry_res = open_telemetry_server_future => {
                match open_telemetry_res {
                    Ok(open_telemetry_res) => open_telemetry_res.context(UnableToStartOpenTelemetryServerSnafu),
                    Err(source) => {
                        Err(Error::UnableToJoinTask { source })
                    }
                }
            },
            pods_watcher_res = async {
                if let Some(fut) = pods_watcher_future {
                    fut.await
                } else {
                    futures::future::pending().await
                }
            } => {
                pods_watcher_res.context(UnableToInitializePodsWatcherSnafu)
            },
            () = shutdown_signal() => {
                tracing::info!("Goodbye!");
                Ok(())
            },
        }
    }

    /// Updates all of the component statuses to `Initializing`.
    pub async fn set_components_initializing(&self) {
        let app_lock = self.app.read().await;
        let Some(app) = app_lock.as_ref() else {
            return;
        };

        let valid_datasets = Self::get_valid_datasets(app, LogErrors(false));
        for ds in &valid_datasets {
            self.status
                .update_dataset(&ds.name, ComponentStatus::Initializing);
        }

        if cfg!(feature = "models") {
            for embedding in &app.embeddings {
                self.status
                    .update_embedding(&embedding.name, ComponentStatus::Initializing);
            }

            for model in &app.models {
                self.status
                    .update_model(&model.name, ComponentStatus::Initializing);
            }

            for tool in &app.tools {
                self.status
                    .update_tool(&tool.name, ComponentStatus::Initializing);
            }

            for tool_catalog in default_available_catalogs() {
                self.status
                    .update_tool_catalog(tool_catalog.name(), ComponentStatus::Initializing);
            }
        }

        let valid_catalogs = Self::get_valid_catalogs(app, LogErrors(false));
        for catalog in valid_catalogs {
            self.status
                .update_catalog(&catalog.name, ComponentStatus::Initializing);
        }

        let valid_views = Self::get_valid_views(app, LogErrors(false));
        for view in valid_views {
            self.status
                .update_view(&view.name, ComponentStatus::Initializing);
        }
    }

    /// Will load all of the components of the Runtime, including `secret_stores`, `catalogs`, `datasets`, `models`, and `embeddings`.
    ///
    /// The future returned by this function will not resolve until all components have been loaded and marked as ready.
    /// This includes waiting for the first refresh of any accelerated tables to complete.
    pub async fn load_components(&self) {
        self.set_components_initializing().await;

        self.start_extensions().await;

        // Must be loaded before datasets
        self.load_embeddings().await;

        // Spawn each component load in its own task to run in parallel
        let task_history = tokio::spawn({
            let self_clone = self.clone();
            async move {
                if let Err(err) = self_clone.init_task_history().await {
                    tracing::warn!("Creating internal task history table: {err}");
                }
            }
        });

        let results_cache = tokio::spawn({
            let self_clone = self.clone();
            async move {
                self_clone.init_results_cache().await;
            }
        });

        let datasets = tokio::spawn({
            let self_clone = Arc::new(self.clone());
            async move {
                self_clone.load_datasets().await;
            }
        });

        let catalogs = tokio::spawn({
            let self_clone = self.clone();
            async move {
                self_clone.load_catalogs().await;
            }
        });

        let models = tokio::spawn({
            let self_clone = self.clone();
            async move {
                self_clone.load_models().await;
            }
        });

        let eval_scorer = tokio::spawn({
            let self_clone = self.clone();
            async move {
                let app_lock = self_clone.app.read().await;

                if !cfg!(feature = "models")
                    && app_lock.as_ref().is_some_and(|s| !s.evals.is_empty())
                {
                    tracing::error!("Cannot load evals without the 'models' feature enabled. {ENABLE_MODEL_SUPPORT_MESSAGE}");
                }

                #[cfg(feature = "models")]
                {
                    self_clone.load_eval_scorer().await;
                    let an_eval_exists = app_lock.as_ref().is_some_and(|app| !app.evals.is_empty());
                    if !an_eval_exists {
                        tracing::trace!("No eval spice components defined. Therefore not loading eval tables into database.");
                    } else if let Err(err) = self_clone.load_eval_tables().await {
                        tracing::warn!("Creating internal eval run table: {err}");
                    }
                }
            }
        });

        // Wait for all tasks to complete
        let load_result = tokio::try_join!(
            task_history,
            results_cache,
            datasets,
            catalogs,
            models,
            eval_scorer
        );

        if let Err(err) = load_result {
            tracing::error!("Could not start the Spice runtime: {err}");
        }
    }

    // Closes and deallocates all resources (including the static registries)
    pub async fn close(self) {
        dataconnector::unregister_all().await;
        catalogconnector::unregister_all().await;
        dataaccelerator::unregister_all().await;
        tools::factory::unregister_all_factories().await;
        document_parse::unregister_all().await;
        self.df.shutdown().await;
    }
}

#[allow(clippy::implicit_hasher)]
pub async fn get_params_with_secrets(
    secrets: Arc<RwLock<Secrets>>,
    params: &HashMap<String, String>,
) -> HashMap<String, SecretString> {
    let secrets = secrets.read().await;

    let mut params_with_secrets: HashMap<String, SecretString> = HashMap::new();

    // Inject secrets from the user-supplied params.
    // This will replace any instances of `${ store:key }` with the actual secret value.
    for (k, v) in params {
        let secret = secrets.inject_secrets(k, ParamStr(v)).await;
        params_with_secrets.insert(k.clone(), secret);
    }

    params_with_secrets
}

#[must_use]
pub fn spice_data_base_path() -> String {
    let Ok(working_dir) = std::env::current_dir() else {
        return ".".to_string();
    };

    let base_folder = working_dir.join(".spice/data");
    base_folder.to_str().unwrap_or(".").to_string()
}

pub(crate) fn make_spice_data_directory() -> Result<()> {
    let base_folder = spice_data_base_path();
    std::fs::create_dir_all(base_folder).context(UnableToCreateDirectorySnafu)
}
