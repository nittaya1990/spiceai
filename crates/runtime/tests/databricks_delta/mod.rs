/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::Arc;

use app::AppBuilder;

use crate::{
    get_test_datafusion, init_tracing, run_query_and_check_results, utils::test_request_context,
    ValidateFn,
};
use runtime::{status, Runtime};
use spicepod::component::{dataset::Dataset, params::Params};

fn make_dataset(path: &str, name: &str) -> Dataset {
    let mut dataset = Dataset::new(format!("databricks:{path}"), name.to_string());
    dataset.params = Some(get_params());
    dataset
}

fn get_params() -> Params {
    Params::from_string_map(
        vec![
            (
                "databricks_endpoint".to_string(),
                "${ env:DATABRICKS_HOST }".to_string(),
            ),
            (
                "databricks_token".to_string(),
                "${ env:DATABRICKS_TOKEN }".to_string(),
            ),
            (
                "databricks_aws_secret_access_key".to_string(),
                "${ env:AWS_DATABRICKS_DELTA_SECRET_ACCESS_KEY }".to_string(),
            ),
            (
                "databricks_aws_access_key_id".to_string(),
                "${ env:AWS_DATABRICKS_DELTA_ACCESS_KEY_ID }".to_string(),
            ),
            ("client_timeout".to_string(), "120s".to_string()),
            ("mode".to_string(), "delta_lake".to_string()),
        ]
        .into_iter()
        .collect(),
    )
}

#[tokio::test]
async fn databricks_delta_lake_integration_test() -> Result<(), anyhow::Error> {
    type QueryTests<'a> = Vec<(&'a str, &'a str, Option<Box<ValidateFn>>)>;
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let app = AppBuilder::new("databricks_delta_lake_connector")
                .with_dataset(make_dataset(
                    "spiceai_sandbox.integration.delta_all_types_table",
                    "datatypes",
                ))
                .build();

            let status = status::RuntimeStatus::new();
            let df = get_test_datafusion(Arc::clone(&status));

            let mut rt = Runtime::builder()
                .with_datafusion(df)
                .with_app(app)
                .build()
                .await;

            // Set a timeout for the test
            tokio::select! {
                () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
                    return Err(anyhow::anyhow!("Timed out waiting for datasets to load"));
                }
                () = rt.load_components() => {}
            }

            let queries: QueryTests = vec![(
                "SELECT * FROM datatypes",
                "select",
                Some(Box::new(|result_batches| {
                    for batch in &result_batches {
                        assert_eq!(batch.num_columns(), 15, "num_cols: {}", batch.num_columns());
                        assert_eq!(batch.num_rows(), 1, "num_rows: {}", batch.num_rows());
                    }

                    // snapshot the values of the results
                    let results = arrow::util::pretty::pretty_format_batches(&result_batches)
                        .expect("should pretty print result batch");
                    insta::with_settings!({
                        description => format!("Databricks (mode: delta_lake) Integration Test Results"),
                        omit_expression => true,
                        snapshot_path => "../snapshots"
                    }, {
                        insta::assert_snapshot!(format!("databricks_delta_lake_select"), results);
                    });
                })),
            )];

            for (query, snapshot_suffix, validate_result) in queries {
                run_query_and_check_results(
                    &mut rt,
                    &format!("databricks_delta_lake_test_{snapshot_suffix}"),
                    query,
                    true,
                    validate_result,
                )
                .await
                .map_err(|e| anyhow::anyhow!("{e}"))?;
            }

            Ok(())
        })
        .await
}
