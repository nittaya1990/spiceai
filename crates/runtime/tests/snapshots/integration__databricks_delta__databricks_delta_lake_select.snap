---
source: crates/runtime/tests/databricks_delta/mod.rs
description: "Databricks (mode: delta_lake) Integration Test Results"
---
+----------------+-----------------+-----------+-----------+----------+-----------+------------+-------------+------------------------------+------------+----------------------+-------------+-----------+-------------------------------+--------------------+
| string_col     | long_col        | int_col   | short_col | byte_col | float_col | double_col | boolean_col | binary_col                   | date_col   | timestamp_col        | decimal_col | array_col | struct_col                    | map_col            |
+----------------+-----------------+-----------+-----------+----------+-----------+------------+-------------+------------------------------+------------+----------------------+-------------+-----------+-------------------------------+--------------------+
| example_string | 123456789012345 | 123456789 | 12345     | 123      | 123.45    | 1234567.89 | true        | 6578616d706c655f62696e617279 | 2024-12-04 | 2024-12-04T12:34:56Z | 123456.78   | [1, 2, 3] | {id: 1, name: example_struct} | {key1: 1, key2: 2} |
+----------------+-----------------+-----------+-----------+----------+-----------+------------+-------------+------------------------------+------------+----------------------+-------------+-----------+-------------------------------+--------------------+
